include/internal/tsan_assist.h

```diff
@@ -28,12 +29,22 @@
  * bother. Having Thread Sanitizer accept "thread-opportunistic" code
  * allows to move on trouble-shooting real bugs.
  *
- * We utilize the fact that compilers that implement Thread Sanitizer
- * implement even atomic operations. Then it's assumed that
- * ATOMIC_{LONG|INT}_LOCK_FREE are assigned same value as
+ * Resolving Thread Sanitizer nits was the initial purpose for this module,
+ * but it was later extended with more nuanced primitives that are useful
+ * even in "non-opportunistic" scenarios. Most notably verifying if a shared
+ * structure is fully initialized and bypassing the initialization lock.
+ * It's suggested to view macros defined in this module as "annotations" for
+ * thread-safe lock-free code, "Thread-Safe ANnotations"...
+ *
+ * It's assumed that ATOMIC_{LONG|INT}_LOCK_FREE are assigned same value as
  * ATOMIC_POINTER_LOCK_FREE. And check for >= 2 ensures that correspodning
  * code is inlined. It should be noted that statistics counters become
  * accurate in such case.
+ *
+ * Special note about TSAN_QUALIFIER. It might be undesired to use it in
+ * a shared header. Because whether operation on specific variable or member
+ * is atomic or not might be irrelevant in other modules. In such case one
+ * can use TSAN_QUALIFIER in cast specifically when it has to count.
```

Comment:
On the other hand, I think the compiler will give a warning/error if you try to use an _Atomic variable in a non-atomic way.

Reply:
What do you mean? Compiler issues warning if you take pointer to atomic variable that is then implicitly casted to non-atomic(*), but if variable or pointer is declared _Atomic it will just generate corresponding code...

(*) like in
```
void foo(int *p);
void bar() { _Atomic int i; foo(&i); }
```